Add client demo and some basic client side features

emdashcodes opened the pull request:
See `AI-135/create-client-side-features`

## Description

This PR accomplishes two main goals:

1. Enables testing of client-side features within the demo chat application by implementing the basic structure for the agent to request client-side execution and process results.
2. Introduces an initial set of core client features (`src/client/features/`):

   - A generic navigation feature (`core/navigate`)
   - Block manipulation features (`core/insert-paragraph-block`, `core/insert-heading-block`) for adding content

This PR includes a few areas that could benefit from additional cleanup or future improvements:

- Some of the feature execution logic currently lives in the demo. Ideally, this could be abstracted into the SDK. It’s not yet clear where the line should be between what Agent/Client developers should implement and what the SDK should handle out of the box since we need the agent to tell us to execute the client feature and then we need to somehow send that response back.
- See https://github.com/Automattic/wp-feature-api/pull/19 where I did further cleanup around the packages. ~Potentially breaking things down into separate packages like `core-feature-api` and `core-features-library`, per the comment here: https://github.com/Automattic/wp-feature-api/pull/9#issuecomment-2769375643. Since the PR was already getting big, I held off on setting up a full monorepo structure, but that could probably also help clean up the wp.data workaround.~

### Key Changes

- Implemented the client feature request/response flow in `BasicAgent` and `ChatController`
- Added core client features (`navigate`, `insert-paragraph`, `insert-heading`) in the SDK (`src/client/features/`)
- Implemented client feature registration in the SDK (`src/client/features/index.ts`)
- Added logic to skip sending results for certain features like navigation (based on `output_schema`)

## Testing Instructions

1. Ensure the demo plugin is active and an OpenAI API key is configured.
2. Build both the main plugin and the demo plugin (`npm install && npm run build`).
3. Load a WordPress admin page (e.g., Dashboard).
4. Open the demo chat UI.
5. Ask the agent to perform actions requiring the new client features:
    - "Navigate to the General Settings page." (Verify the browser navigates to `wp-admin/options-general.php`)
    - "Navigate to the new post editor." (Verify the browser navigates to `wp-admin/post-new.php`)
    - "Add a paragraph block with the text 'Hello World'." (Verify a paragraph block is added to the editor—if open—and a "Feature Call Result" appears in the chat)
    - "Insert a heading that says 'My New Section'." (Verify a heading block is added and a "Feature Call Result" appears)
    - "Add a heading level 3 that says 'Subsection'." (Verify a level 3 heading is added)
6. Test server-side features (e.g., "Show site info") to ensure they still work and return results correctly.

## Demo

https://github.com/user-attachments/assets/4fb5a1ac-71d0-4b4a-8e21-767ac4d78fd7

emdashcodes commented on tools/eslint/import-resolver.js:
```diff
@@ -15,17 +11,5 @@ exports.resolve = function ( source, file, config ) {
 			extensions: [ '.tsx', '.ts', '.mjs', '.js', '.json', '.node' ],
 		} );
 
```
This was causing import / linter issues for me, and removing it seemed to resolve them. Is this still needed?

jorgefilipecosta replied:
This was part of the gutenberg setup which we tried mimic, but if it causing issues I guess it could be removed.

emdashcodes commented on demo/wp-feature-api-demo/src/chat-app.tsx:
```diff
@@ -124,6 +108,7 @@ const ChatAppContent = () => {
 						icon={ isLoading ? null : arrowRight }
 					>
 						{ isLoading ? (
+							// @ts-expect-error: Spinner is not returning the correct type.
 							<Spinner />
```
I'm not entirely sure why, but I was getting type errors even though everything points to `<Spinner />` being valid.

emdashcodes commented on src/client/types.ts:
```diff
@@ -0,0 +1,22 @@
+export interface Feature {
```
These types were referenced used in the original client side PR (https://github.com/Automattic/wp-feature-api/pull/9), but this file did not seem to exist.

emdashcodes commented on demo/wp-feature-api-demo/src/hooks/use-client-action-handler.ts:
```diff
@@ -0,0 +1,167 @@
+	const executeClientFeature = useCallback(
+		async ( featureId: string, args: any ): Promise< unknown > => {
+			// Access store via global wp.data
+			// @todo We should probably export our feature store as a proper package and use that instead,
```
This is handled in https://github.com/Automattic/wp-feature-api/pull/19

emdashcodes commented on demo/wp-feature-api-demo/src/hooks/use-message-handler.ts:
```diff
@@ -0,0 +1,115 @@
+
+			try {
+				// Access store via global wp.data
+				// @todo We should probably export our feature store as a proper package and use that instead,
```
This is cleaned up in https://github.com/Automattic/wp-feature-api/pull/19

jorgefilipecosta commented on demo/wp-feature-api-demo/includes/Agent/BasicAgent.php:
```diff
@@ -119,23 +169,130 @@ private function make_tool_call(): CreateResponseMessage {
+	 * @param array<WP_Feature> $features The server features to convert.
+	 * @return array<array{type: string, function: array}> The tools for the LLM.
+	 */
+	private function tools_from_server_features( array $features ): array {
```
We have an issue causing problems for MCP where we mark all arguments as required I think because of the way the openAI API works. I guess that transformation should be done here.
(Not related to this PR just a note for the future).

jorgefilipecosta commented on demo/wp-feature-api-demo/src/hooks/use-client-action-handler.ts:
```diff
@@ -0,0 +1,167 @@
+			// Access store via global wp.data
+			// @todo We should probably export our feature store as a proper package and use that instead,
+			// this is to avoid double bundling by calling things from the root src/client/... files directly
+			const selector = ( window.wp as any )?.data?.select(
```
We could avoid having the window.wp global usage. For example by importing the useRegistry:
import { useRegistry } from '@wordpress/data';
And then doing registry.select...

emdashcodes replied:
This was a temporary workaround for some issues I was having accessing the store properly, but this gets fixed in my follow-up PR: https://github.com/Automattic/wp-feature-api/pull/19/files#diff-2ad67c14aab9bf8b2de50ebdd1ef4c360545c084739e2b0ed2405c7adb84d140

jorgefilipecosta commented on demo/wp-feature-api-demo/src/hooks/use-client-action-handler.ts:
```diff
@@ -0,0 +1,167 @@
+
+			const { id: featureId, args, tool_call_id: toolCallId } = action;
+
+			const selector = ( window.wp as any )?.data?.select(
```
Same as above

jorgefilipecosta commented on src/client/features/blocks.ts:
```diff
@@ -0,0 +1,138 @@
+				description: __( 'Text content for the paragraph.' ),
+			},
+		},
+		required: [ 'content' ],
```
This is a good case to test optional parameters, it should be possible to insert empty paragraphs, and it is but the assistant passes content with an empty string. Would it be possible to simply make content not a required property on the definition? When passing tools to the OpenAi we can adapt but ideally the definitions don't force all arguments as required.

emdashcodes replied:
I agree with this. I believe the other blocks (like heading) should also accept optional parameter for size and have a default.

I currently have it marked as required due to `all_fields_required` is(https://a8c.slack.com/archives/C08J7H30TNH/p1744390987207359?thread_ts=1744390495.325079&cid=C08J7H30TNH), and haven't investigated that further -- but I'll look at that soon.

emdashcodes replied:
Okay, I see what I was missing -- the client side definitions needed to be run through the schema adapter. I've updated `content` to be optional for the paragraph block, and removed `level` from being required on the heading block.

https://github.com/Automattic/wp-feature-api/pull/18/commits/c2d4207394a0c605a39306ecfb5c2a90f1f3e72e

jorgefilipecosta commented on src/client/features/index.ts:
```diff
@@ -0,0 +1,29 @@
+export const registerDefaultWPFeatures = () => {
+	const { registerFeature } = dispatch( featureStore );
+
+	if ( navigate ) {
```
Minor could be set as
[ navigate, insertParagraphBlock, insertHeadingBlock ].filter( (feature) ==> !! feature).forEach( registerFeature );

emdashcodes replied:
Updated in https://github.com/Automattic/wp-feature-api/pull/18/commits/c2d4207394a0c605a39306ecfb5c2a90f1f3e72e

jorgefilipecosta commented on src/client/features/navigation.ts:
```diff
@@ -0,0 +1,83 @@
+		},
+		required: [ 'url' ],
+	},
+	callback: ( args: { url: string } ) => {
```
Minor: The logic on this function has an issue:
My dev site is at http://localhost:6888/site-wp-dev/.
I said "Navigate to the posts list" the tool invoked was:
{
  "url": "/wp-admin/edit.php"
}
Which seems ok, but the redirect was to http://localhost:6888/wp-admin/edit.php instead of http://localhost:6888/site-wp-dev/wp-admin/edit.php.

emdashcodes replied:
Fixed in https://github.com/Automattic/wp-feature-api/pull/18/commits/9bb6cca4f05548b781dbe06dd665722d0c23bef0

jorgefilipecosta reviewed (commented):
When I use this prompt:
"Insert info about my site as a paragraph"
I get an error:
<img width="1460" alt="Screenshot 2025-04-14 at 16 52 15" src="https://github.com/user-attachments/assets/83c1b319-e211-4494-adbc-d410ff43b016" />

I guess we are not handling cases where the output of a tool is the input of other.

jorgefilipecosta reviewed (commented):
This is not related to this PR, as it follows the existing pattern. However, it seems we might be leaking information about how WordPress features work into the agent code (e.g., the distinction between client and server features). This could be adding unnecessary complexity to the code.
Here's how I envision it:

The only interaction between the agent code and WordPress should be through:
- wp.data.select('feature-api').getRegisteredFeatures(); to discover features. This discovery could be dynamic. Depending on the user's context (where they are in the application), we might show different features. Querying features might also involve server-side calls with logic to determine relevant features, but this logic should be independent of the agent code.
- wp.data.select('feature-api').getRegisteredFeatureCallback(); to get the callback function for a specific feature.

The agent code shouldn't have any other interactions with WordPress. There should be no distinction between server and client features from the agent's perspective. If the information for a feature resides on the server, the feature's callback should automatically handle the request to the server, without the agent needing to be aware of it.
Ideally, the agent code should be totally generic and independent of WordPress. It would simply receive an array of features and a callback to execute a chosen feature. This agent could potentially work with WordPress, Drupal, or any other underlying system.
Think of the agent code as an in-browser MCP client, where the tools and resources are provided by wp.data.select('feature-api').getRegisteredFeatures();.
The agent code could be written in JS/TS (proxying third-party API calls through PHP to avoid leaking API keys), or it could be written in PHP. If written in PHP, this PHP code should remain independent of WordPress specifics and simply forward messages and tool information to the relevant third-party SDKs.

This is off course not a blocker to this PR, just sharing what I think could be the end result.

jorgefilipecosta reviewed (approved):
Left some enhancements that could be done before the merge, but things tested well and I was able to execute client side features with success from the agent, so I think it ready to merge.

emdashcodes commented:
Thanks for the review and testing! I've addressed some of the feedback here. The rest is either handled in https://github.com/Automattic/wp-feature-api/pull/19 or will be improved with a cleaner agent approach.

Pasting my [Slack reply to the above](https://a8c.slack.com/archives/C08J7H30TNH/p1744649127016549?thread_ts=1744370036.817069&cid=C08J7H30TNH) for completeness:

> I totally agree with your comment there. I had some concerns about that too (and was a bit worried too much of the execution logic ended up coupled in the demo agent). But I was also worried about changing too much while I was still getting familiar last week. :sweat_smile:
I think a cleaner, more decoupled agent would make for a stronger demo and a better example for folks looking in our repo or looking to build on top of the Feature API. What I’d suggest is getting my current chain of PRs merged (gives a basic demo), and then starting a fresh agent implementation based on your feedback.
I was already going to look this week at adding to the demo to support things like tool call chaining and message history, and that might be easier to do with a cleaner foundation :simple_smile:

I checked with James too, and he agrees it makes sense to create a cleaner demo agent.
